### PR TITLE
refactor(@chat): add chat members

### LINKED ISSUE #4611: [base_bc] making a member as admin of group chat does not hide the context menu for this member
### Description

image

### LINKED ISSUE #4613: [base_bc] can't remove a member from a group
### Description

```
ERR 2022-01-28 10:18:44.295+03:00 error while removing member from group:    topics="chat-service" tid=4102993 file=service.nim:388 msg="\nstatus-go error [methodName:wakuext_removeMemberFromGroupChat, code:-32000, message:invalid event protocol.MembershipUpdateEvent{Type:5, ClockValue:0x17e9f8cfda5, Members:[]string{\"0x04bed4a6ce4a8c3bf06573fd4d1d5b17d83410b09861773ded129fae485d63a0c9264d806d2f17524cef33a5dc14050c211024e18daaadfa45010d8a287b3198b8\"}, Name:\"\", From:\"0x043067fb0a2374b44d41800a546b26105bff33d4654f7d4e8dc34744d1f6af7cf38b380b4d1fb7a78bef6a90845b399c91b348d75431bf0103f04ba7dc18fd2202\", Signature:[]uint8{0x54, 0x7b, 0xed, 0x19, 0xef, 0xf2, 0xcc, 0xf7, 0xdb, 0x8a, 0x2, 0x84, 0xe3, 0x2f, 0x3e, 0x2e, 0x31, 0xa5, 0x68, 0xf, 0xe2, 0x65, 0xbc, 0xf6, 0x7, 0x60, 0xd7, 0xe4, 0xa5, 0x97, 0xe5, 0x39, 0x29, 0x42, 0x24, 0x65, 0xcf, 0x2d, 0xb6, 0xac, 0xaf, 0x78, 0x38, 0xbb, 0x9e, 0x46, 0x99, 0xd, 0x0, 0x5c, 0x82, 0x9e, 0xef, 0xe6, 0x79, 0xed, 0x7c, 0xc1, 0x24, 0x9b, 0xd0, 0x7b, 0xfd, 0x4, 0x1}, ChatID:\"cc3c0605-a871-4f07-8954-f8933efd53cd-0x043067fb0a2374b44d41800a546b26105bff33d4654f7d4e8dc34744d1f6af7cf38b380b4d1fb7a78bef6a90845b399c91b348d75431bf0103f04ba7dc18fd2202\", RawPayload:[]uint8{0x8, 0xa5, 0xfb, 0xb3, 0xfc, 0xe9, 0x2f, 0x12, 0x84, 0x1, 0x30, 0x78, 0x30, 0x34, 0x62, 0x65, 0x64, 0x34, 0x61, 0x36, 0x63, 0x65, 0x34, 0x61, 0x38, 0x63, 0x33, 0x62, 0x66, 0x30, 0x36, 0x35, 0x37, 0x33, 0x66, 0x64, 0x34, 0x64, 0x31, 0x64, 0x35, 0x62, 0x31, 0x37, 0x64, 0x38, 0x33, 0x34, 0x31, 0x30, 0x62, 0x30, 0x39, 0x38, 0x36, 0x31, 0x37, 0x37, 0x33, 0x64, 0x65, 0x64, 0x31, 0x32, 0x39, 0x66, 0x61, 0x65, 0x34, 0x38, 0x35, 0x64, 0x36, 0x33, 0x61, 0x30, 0x63, 0x39, 0x32, 0x36, 0x34, 0x64, 0x38, 0x30, 0x36, 0x64, 0x32, 0x66, 0x31, 0x37, 0x35, 0x32, 0x34, 0x63, 0x65, 0x66, 0x33, 0x33, 0x61, 0x35, 0x64, 0x63, 0x31, 0x34, 0x30, 0x35, 0x30, 0x63, 0x32, 0x31, 0x31, 0x30, 0x32, 0x34, 0x65, 0x31, 0x38, 0x64, 0x61, 0x61, 0x61, 0x64, 0x66, 0x61, 0x34, 0x35, 0x30, 0x31, 0x30, 0x64, 0x38, 0x61, 0x32, 0x38, 0x37, 0x62, 0x33, 0x31, 0x39, 0x38, 0x62, 0x38, 0x20, 0x5}} ]\n"
```

Screenshot 2022-01-28 at 10 27 02

### PR DESCRIPTION
**Pull request by alaibe**
fixes https://github.com/status-im/status-desktop/issues/4612
fixes https://github.com/status-im/status-desktop/issues/4613
fixes #4611

### CODE REVIEW

**Review by saledjenic — COMMENTED** [review]
@alaibe on the first sight code itself looks fine (didn't go through details yet, just briefly took a look, will check in details later, but let's kick off the discussion about it firstly) and as I see it's almost the same/similar to what we have for `users module`, so I am wondering why wouldn't you just update that module (added fields you need) and use it here for that purpose as well cause both parts refers to the list of members of a certain chat.

Current user module props:
```
ModelRole {.pure.} = enum
    Id = UserRole + 1
    Name
    OnlineStatus
    Icon
    IsIdenticon
```

props in chat members module:
```
  ModelRole {.pure.} = enum
    Id = UserRole + 1
    Name
    Icon
    IsIdenticon
    IsAdmin
    Joined
```

honestly I don't see the difference between chat members and users, as they both will have in each moment the same list of users. Even maybe in some moment in future we decide to display user online status in `GroupInfoPopup`.

Also I don't see that this fixes the issue we discussed there when someone left the chat. Seems that is not handled here as well.

**Comment by alaibe** [comment]
@saledjenic at first sight I thought like you, then here is what I found:
- There is different attributes in both list. For public chat or 1-1 chat there is no admin or joined boolean.

Then the content of the list itself are different:
- For communities there is not chat members, as this is managed at communities level
- In a public chat, you are in the user list as soon as you send a message, so imagine this scenario:

Create a group with 2 members A and B.
B send a message to the group and
A remove B from the group

According to user list rule A will be present in it but he should not be (This is the bug I am fixing)
(This is just an example there is other cases)

Another option would have been to do check `if chat.is_group_chat` in most of the function. I found the solution of adding a dedicated module more easy to manage

Hope that makes more sense

**Comment by saledjenic** [comment]
> There is different attributes in both list. For public chat or 1-1 chat there is no admin or joined boolean

I agree that `IsAdmin` and `Joined` fields are not applicable in case public or 1:1 chat, but doesn't matter just make them `false` in such cases. The thing is that chat members module and users module are actually the same in their nature. List used to display chat members and list used to display list of users on the right will always be the same. That's actually 2 views to the same source.

Why' wouldn't we have just 2 fields more to the users model instead of introducing new module (that's heavy change which carries lot of duplications since it needs to respond to the same signals as users module which is already in place):
```
ModelRole {.pure.} = enum
    Id = UserRole + 1
    Name
    OnlineStatus
    Icon
    IsIdenticon
    IsAdmin
    Joined
```

> For communities there is not chat members, as this is managed at communities level

I agree there are no chat members in case of community, but in that case according to the changes applied here entire module is used nothing for, it will be still here, but not displayed (kind of overhead). That's actually one more advantage of sharing users model (in case the chat type is community, we won't have an option for the chat list, so users model is shared in that case, both places point to the same list). 

> Create a group with 2 members A and B.
B send a message to the group and
A remove B from the group
>
> According to user list rule A will be present in it but he should not be (This is the bug I am fixing)
(This is just an example there is other cases)

Currently it will be still in the list, but the reason is simple, cause that is a bug. I checked how the Discord handles this, so when user leaves the group chat, he should be removed from the users list + chat list (in our case), but the messages remain. So this case will be simply handled connecting `users module` to one signal more. What confirms that our curren `users module` and module added here are same in their nature.

My suggestion is to revert this changes and add 2 necessary fields to the users module.

Lets involve more people and make a consensus on this, till then I will request a changes on this PR.

**Review by saledjenic — CHANGES_REQUESTED** [review]
@alaibe I am requesting a changes on this PR, till we are all on the same page about this. Last comment [here](https://github.com/status-im/status-desktop/pull/4661#issuecomment-1027913931).

**Comment by jrainville** [comment]
I read though the comments and checked the code.

On first look, it does seem like this new module/model is a duplication of the usersModule.
I do see that in the case of the group chats vs public chats, the membersModule and the usersModule seem to have both a case (members for group and users for public).
However, I think that's just a problem with the way to usersModule is implemented right now and it can be improved to fix the "user leaves the group" use case so that in the case of a group chat, the usersModule accurately represents the current members.

If the approach above incurs too many `if/else`s, then maybe what we can do is have a step in the middle where we do have a separate model for members, but it can still use the same Model class, by just adding the missing `joined` and `isAdmin` properties to the Item, like Sale proposed.

So, in summary, I think we could try to reuse the same code as much as we can. The problems faced with the usersModule as it is right now should be fixed if possible.

**Comment by alaibe** [comment]
ok will update

**Comment by osmaczko** [comment]
> In a public chat, you are in the user list as soon as you send a message

This should not be the case anymore. Public chat should not have a user list.
In the case of group chats, the list of users should not be built based on the messages sent, as some participants may have already left.

@iurimatias please confirm

**Comment by saledjenic** [comment]
> ok will update

thank you

**Comment by alaibe** [comment]
@osmaczko  
> This should not be the case anymore. Public chat should not have a user list.

They don t have a user list in the backend because it is only subscription but we are building one in the client manually based on message send

> In the case of group chats, the list of users should not be built based on the messages sent, as some participants may have already left.

I agree this is the change being discussed

**Comment by alaibe** [comment]
@saledjenic @jrainville code updated

File: src/app/modules/main/chat_section/chat_content/users/module.nim [hunk 1]
```diff
@@ -45,27 +50,55 @@ method isLoaded*(self: Module): bool =
 
 method viewDidLoad*(self: Module) =
   # add me as the first user to the list
+  var admin = false
```
Comment (saledjenic) [inline src/app/modules/main/chat_section/chat_content/users/module.nim:53]: @alaibe all changes done here in `viewDidLoad` may be easily reverted and changed with the changes I did [here](https://github.com/status-im/status-desktop/pull/4649/files) in `getMembersPublicKeys` and on top of that because of added `isAdmin` and `joined` props for `item`, you just need to update `initItem` in `viewDidLoad`. So less code and easier for understanding without a lot of "if".

File: src/app_service/service/chat/service.nim [hunk 2]
```diff
@@ -70,6 +76,7 @@ const SIGNAL_CHAT_HISTORY_CLEARED* = "chatHistoryCleared"
 const SIGNAL_CHAT_RENAMED* = "chatRenamed"
 const SIGNAL_CHAT_MEMBERS_ADDED* = "chatMemberAdded"
 const SIGNAL_CHAT_MEMBER_REMOVED* = "chatMemberRemoved"
+const SIGNAL_CHAT_MEMBER_UPDATED* = "chatMemberUpdated"
```
Comment (saledjenic) [inline src/app_service/service/chat/service.nim:79]: @alaibe the issue when someone left a group chat is till here. If you create a chat and add members and if any of them left the chat members list (as well as members count) doesn't get updated. For that we have to monitor on `messages.new` signal and check `members` list of concrete chat received as part of the `chats` array and update the list.

**Review by saledjenic — CHANGES_REQUESTED** [review]
@alaibe apart of 2 issues I added below I am suggesting if you agree to get [this](https://github.com/status-im/status-desktop/pull/4649) merged firstly and then you can rebase/update this PR accordingly.

File: src/app_service/service/chat/service.nim (hunk 2 above)
Comment (alaibe) [reply src/app_service/service/chat/service.nim:79]: Yep, but this is not the issue I am fixing, I am fixing the issue of when you remove a member from a group he is not in the group list anymore. What you describe is the opposite, which I haven't test and is probably what you describe. Can we have a separate issue for that. This one is already fixing 3 issues

File: src/app/modules/main/chat_section/chat_content/users/module.nim (hunk 1 above)
Comment (alaibe) [reply src/app/modules/main/chat_section/chat_content/users/module.nim:53]: as u prefer

File: src/app/modules/shared_models/user_model.nim [hunk 3]
```diff
@@ -140,14 +148,19 @@ QtObject:
...
       return
 
     self.items[ind].name = name
     self.items[ind].icon = icon
     self.items[ind].isIdenticon = isIdenticon
+    self.items[ind].isAdmin = isAdmin
+    self.items[ind].joined = joined
     
     let index = self.createIndex(ind, 0, nil)
     self.dataChanged(index, index, @[ModelRole.Name.int, ModelRole.Icon.int, ModelRole.IsIdenticon.int])
```
Comment (saledjenic) [inline src/app/modules/shared_models/user_model.nim:166]: @alaibe one minor thing remained. Please either add `ModelRole. IsAdmin.int, ModelRole. Joined.int` roles to the list of changed roles of `dataChanged` signal or add new proc here to `user_model.nim` similar to one we already have but for updating other params like this `proc updateItem*(self: Model, id: string, isAdmin: bool = false, joined: bool = false) =` (I think that proc name overloading this way is allowed in Nim) and then in the `chat_section/chat_content/users/module.nim`, method `onChatMemberUpdated` will update only fields which actually were changed:
```
method onChatMemberUpdated*(self: Module, publicKey: string, admin: bool, joined: bool) =
  self.view.model().updateItem(publicKey, admin, joined)
```

**Review by saledjenic — APPROVED** [review]
@alaibe changes look good, if you don't mind, please just do the final update in the way you prefer the most following a comment below.

File: src/app/modules/main/chat_section/chat_content/users/module.nim [hunk 4]
```diff
@@ -47,27 +50,41 @@ method isLoaded*(self: Module): bool =
...
   self.moduleLoaded = true
   self.delegate.usersDidLoad()
 
 method getModuleAsVariant*(self: Module): QVariant =
   return self.viewVariant
 
-method newMessagesLoaded*(self: Module, messages: seq[MessageDto]) = 
+method newMessagesLoaded*(self: Module, messages: seq[MessageDto]) =
+  let chat = self.controller.getChat()
+  if chat.isPrivateGroupChat():
```
Comment (jrainville) [inline src/app/modules/main/chat_section/chat_content/users/module.nim:85]: We could even make that check to return if it's **not** a public chat? Since it's only public chats that add users on new messages

File: src/app_service/service/chat/service.nim [hunk 5]
```diff
@@ -405,7 +412,15 @@ QtObject:
   method makeAdmin*(self: Service, chatId: string, pubKey: string) =
     try: 
       let response = status_chat.makeAdmin(chatId, pubKey)
-      self.emitUpdate(response)
+      for member in self.chats[chatId].members.mitems:
+        if (member.id == pubKey):
+          member.admin = true
+          echo "doing the emit"
```
Comment (jrainville) [inline src/app_service/service/chat/service.nim:418]: Can be removed

**Review by jrainville — APPROVED** [review]
Looks good. Just get rid of the `echo` please.

Also, take a look at my suggestion, maybe it can save a tiny bit of performance 🤷

File: src/app/modules/main/chat_section/chat_content/users/module.nim (hunk 4 above)
Comment (alaibe) [reply src/app/modules/main/chat_section/chat_content/users/module.nim]: good point, updated

**Review by jrainville — APPROVED** [review]
Approved again 👏

**Review by borismelnik — APPROVED** [review]
